### PR TITLE
mpl/gpu/ze: add engine selection logic to optimize difference copy types (PR #8)

### PR DESCRIPTION
**Pull request by abrooks98**
## Pull Request Description

This PR adds logic for engine selection when utilizing the copy operations in the MPL ZE backend with the goal of optimizing copy performance based on characteristics of the copy (i.e. message size, source/destination location, etc.).

Depends on #6570

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [ ] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

### CODE REVIEW

**Comment by abrooks98** [comment]
Split into smaller commits that are more self-contained

**Comment by abrooks98** [comment]
Moved 7 commits from this PR to #6587 to keep this PR more contained

File: src/mpl/src/gpu/mpl_gpu_ze.c [hunk 1]
```diff
@@ -1596,6 +1596,30 @@ int MPL_gpu_query_pointer_attr(const void *ptr, MPL_pointer_attr_t * attr)
...
+        };
+        ze_device_handle_t device = NULL;
+
+        ret = zeMemGetAllocProperties(ze_context, ptr, &prop, &device);
+        assert(ret == ZE_RESULT_SUCCESS);
+        type = prop.type;
+    } else {
+        type = attr->device_attr.prop.type;
+    }
+    return type != ZE_MEMORY_TYPE_UNKNOWN;
```
Comment (raffenet) [inline src/mpl/src/gpu/mpl_gpu_ze.c:1621]: This comparison in the return statement is confusing. Wouldn't it return true for host memory?
> The application may query attributes of a memory allocation unrelated to the context. When this occurs, the returned allocation type will be [ZE_MEMORY_TYPE_UNKNOWN](https://spec.oneapi.io/level-zero/latest/core/api.html#ze__api_8h_1a6808370b64ade7fdd4a74093224e1d9ba437d57ae7d0652068b62eabdc6cda93a), and the returned identifier and associated device is unspecified.

File: src/mpid/ch4/shm/ipc/src/ipc_send.h [hunk 2]
```diff
@@ -118,7 +118,7 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_IPCI_try_lmt_isend(const void *buf, MPI_Aint
     if (ipc_attr.gpu_attr.type == MPL_GPU_POINTER_DEV) {
         mpi_errno = MPIDI_GPU_get_ipc_attr(mem_addr, rank, comm, &ipc_attr);
         MPIR_ERR_CHECK(mpi_errno);
-    } else {
+    } else if (!MPL_gpu_query_pointer_is_dev(buf, &ipc_attr.gpu_attr)) {
```
Comment (raffenet) [inline src/mpid/ch4/shm/ipc/src/ipc_send.h:121]: Isn't this already checked on line 116 and 118 above? `mem_addr` should point to the start of the data in `buf`.

File: src/mpl/include/mpl_gpu.h [hunk 3]
```diff
@@ -77,10 +77,15 @@ static inline int MPL_gpu_query_pointer_attr(const void *ptr, MPL_pointer_attr_t
     return MPL_SUCCESS;
 }
 
+static inline int MPL_gpu_query_pointer_is_dev(const void *ptr, MPL_pointer_attr_t * attr)
```
Comment (raffenet) [inline src/mpl/include/mpl_gpu.h:80]: Commit title `mpl/gpu: add a new API MPIR_GPU_query_pointer_is_dev` is wrong. The API being added in `MPL_gpu_query_is_pointer_dev`.

File: src/mpid/ch4/shm/ipc/src/ipc_send.h (hunk 2 above)
Comment (abrooks98) [reply src/mpid/ch4/shm/ipc/src/ipc_send.h:121]: The rational behind this is that even host-registered memory (i.e. `zeMemAllocHost`) in ZE is device objects which XPMEM can't properly handle. So this check is to confirm `buf` is regular malloced memory. The other backends don't have this issue (I think), so we just check for `MPL_GPU_POINTER_DEV` in their implementations of `MPL_gpu_query_pointer_is_dev`.

File: src/mpl/src/gpu/mpl_gpu_ze.c (hunk 1 above)
Comment (raffenet) [reply src/mpl/src/gpu/mpl_gpu_ze.c:1621]: Oh man. OK, can we stick a comment here basically saying that?

File: src/mpid/ch4/shm/ipc/src/ipc_send.h (hunk 2 above)
Comment (raffenet) [reply src/mpid/ch4/shm/ipc/src/ipc_send.h:121]: I see. And just to confirm memory from `zeMemAllocHost` doesn't support GPU IPC? I'm wondering if we could instead change the condition on line 118 to
```c
if (MPL_gpu_query_pointer_is_dev(buf, &ipc_attr.gpu_attr)) {
    mpi_errno = MPIDI_GPU_get_ipc_attr(mem_addr, rank, comm, &ipc_attr);
    MPIR_ERR_CHECK(mpi_errno);
}
```
If not, another comment here is probably needed to explain the peculiarity with ZE.

File: src/mpl/include/mpl_gpu.h (hunk 3 above)
Comment (abrooks98) [reply src/mpl/include/mpl_gpu.h:80]: Fixed

File: src/mpid/ch4/shm/ipc/src/ipc_send.h (hunk 2 above)
Comment (abrooks98) [reply src/mpid/ch4/shm/ipc/src/ipc_send.h:121]: I will add a comment about this peculiarity.

I am not fully updated on the current status, but the intention is that GPU IPC will eventually be supported for `zeMemAllocHost`. However, our current cache implementation does not handle this yet. Currently the caches are based on device ID, and host-registered memory does not have a device ID. So we would need to modify the cache to accept this type of memory before enabling it here.

Comment (abrooks98) [reply src/mpid/ch4/shm/ipc/src/ipc_send.h:121]: Added comments and reworded the commit messages slightly to reflect this discussion

**Comment by abrooks98** [comment]
test:mpich/ch4/most
test:mpich/ch4/gpu

File: src/mpid/ch4/shm/ipc/src/ipc_send.h (hunk 2 above)
Comment (raffenet) [reply src/mpid/ch4/shm/ipc/src/ipc_send.h:121]: Looks good. Thanks.

**Review by hzhou — APPROVED** [review]
LGTM

**Comment by abrooks98** [comment]
Not sure why two `cuda/stream` tests timed out. I don't think it is caused by this PR as the behavior of CUDA device buffer paths should be unchanged. The only difference in general for CUDA is taking POSIX fallback path instead of XPMEM for host-registered memory, but this test doesn't hit that case

**Comment by abrooks98** [comment]
test:mpich/ch4/gpu

**Comment by abrooks98** [comment]
Thoughts @hzhou?

**Comment by hzhou** [comment]
> Thoughts @hzhou?

I'll take a look tomorrow.

**Comment by hzhou** [comment]
> Thoughts @hzhou?

@abrooks98 Sorry for the delay. I finally bisected the breaking commits, it is one of mine -- https://github.com/pmodels/mpich/pull/6509/commits/3ac9843d50013cc4f09f10101f9fcc239ad2b1a3. Sorry about that! Anyway, https://github.com/pmodels/mpich/pull/6597 should fix it. I am gonna merge that in a few minutes. Could you rebase and retest?

**Comment by abrooks98** [comment]
> > Thoughts @hzhou?
> 
> @abrooks98 Sorry for the delay. I finally bisected the breaking commits, it is one of mine -- [3ac9843](https://github.com/pmodels/mpich/commit/3ac9843d50013cc4f09f10101f9fcc239ad2b1a3). Sorry about that! Anyway, #6597 should fix it. I am gonna merge that in a few minutes. Could you rebase and retest?

No worries, thanks for finding the issue! Will rebase and retest

**Comment by abrooks98** [comment]
test:mpich/ch4/gpu

**Comment by hzhou** [comment]
I think it is okay to merge. The memory leak in pmi is fixed in #6591

**Pull request merged**